### PR TITLE
Unify ingester and distributor chunks into a single heap merge.

### PR DESCRIPTION
**Pull request by tomwilkie**
Before this change, chunks returned by streaming ingester queries are put into one heap, and chunks from the chunk store another.  These two heaps are then merged using yet another heap; whilst this should have the same algorithmic complexity as using one big heap, in reality is faster to put all the chunks into one heap.

As part of this PR I completely reorganised the querier package, sorry.   I tried to split them out into a separate changeset.

Signed-off-by: Tom Wilkie <tom.wilkie@gmail.com>

### CODE REVIEW

File: pkg/querier/iterators/chunk_merge_iterator.go [hunk 1]
```diff
@@ -22,7 +23,8 @@ type chunkMergeIterator struct {
 	currErr   error
 }
 
-func newChunkMergeIterator(cs []chunk.Chunk) storage.SeriesIterator {
+// NewChunkMergeIterator creats a storate.SeriesIterator for a set of chunks.
```
Comment (cboggs) [inline pkg/querier/iterators/chunk_merge_iterator.go]: Teeny typo nit.

**Comment by leth** [comment]
I read something somewhere yesterday which just prompted this question: is this the kind of situation where timsort would perform well? I.e. does the data have lots of sequential runs?

**Comment by tomwilkie** [comment]
> I read something somewhere yesterday which just prompted this question: is this the kind of situation where timsort would perform well? I.e. does the data have lots of sequential runs?

Thats not really relevant to this PR - we're not changing how we merge/dedupe chunks here, just ensuring that when we use ingester chunk streaming we put them in the same "batch" of chunks as we got from the chunk store, and then merge them.

That being said, its probably worth investigating timsort for chunk merging.  It should perform better than our recursive binary merge strategy, although it will have the downsides of increase memory usage.

**Comment by tomwilkie** [comment]
Finally got round to testing this - for a certain class of queries (~100k chunks, 12hr duration, `sum(container_memory_rss) by (namespace)` this reduced the promqlInnerEval from a reliable 16s to 6s, and the overall query from 20s to 10s.

**Comment by tomwilkie** [comment]
We've had this running in prod now for ~4 days, and the unit tests are now pretty comprehensive.  Going to merge once its green.

**Pull request merged**